File: python/구현/Prob4396.py

""" 실버 4. 지뢰 찾기 """

N = 0           # 지뢰판의 크기
Board = []      # 지뢰판

dr = [-1, -1, 0, 1, 1, 1, 0, -1]
dc = [0, -1, -1, -1, 0, 1, 1, 1]

# --------------------------------------------------------------------------------

def doCountStart(r, c):

    count = 0
    
    for num in range(8):
        
        nr = r + dr[num]
        nc = c + dc[num]

        if nr < 0 or nc < 0 or nr >= N or nc >= N:
            continue
        
        if Board[nr][nc] == '*':
            count += 1

    return count

# --------------------------------------------------------------------------------

N = int(input())                                        # 지뢰판의 크기
Board = [list(input()) for _ in range(N)]               # 지뢰판
Board_clicked = [list(input()) for _ in range(N)]       # 지뢰판

flag = False    # 지뢰 클릭 여부

print(N)

for r in range(N):
    for c in range(N):

        # 클릭 안한 칸
        if Board_clicked[r][c] == '.':
            continue

        # 클릭한 칸
        if Board[r][c] == '.':                          # 지뢰가 없는 칸
            Board_clicked[r][c] = doCountStart(r,c)
        elif not flag and Board[r][c] == '*':            # 지뢰가 있는 칸
            flag = True

if flag:
    for n1 in range(N):
        for n2 in range(N):
                if Board[n1][n2] == '*': 
                    Board_clicked[n1][n2] = '*'

for row in Board_clicked:
    for col in row:
        print(col, end="")
    print()

File: 02 백준/02 구현/Java/src/Prob4396.java

/**
* title  : 4396. 지뢰 찾기 (Silver 4)
* time   : 104ms
* memory : 14140KB
*/

import java.io.*;

public class Prob4396 {

    private static int[][] dir = {{-1, -1}, {-1, 0}, {-1, 1}, 
                                    {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};

    private static int N;
    private static char[][] board;
    private static int[][] answer;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        board = new char[N][N];
        answer = new int[N][N];

        for (int i = 0; i < N; i++) {
            String temp = br.readLine();
            for (int j = 0; j < N; j++) {
                board[i][j] = temp.charAt(j);
                answer[i][j] = -1;
            }
        }

        boolean is_bomb = false;

        for (int i = 0; i < N; i++) {
            char[] temp = br.readLine().toCharArray();
            for (int j = 0; j < N; j++) {
                if (temp[j] == 'x') {
                    if (board[i][j] == '.') {
                        answer[i][j] = count_bomb(i, j);
                    } else if (board[i][j] == '*') {
                        is_bomb = true;
                    }
                }
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (0 <= answer[i][j]) {
                    sb.append(answer[i][j]);
                } else if (is_bomb && board[i][j] == '*'){
                    sb.append('*');
                } else {
                    sb.append('.');
                }
            }
            sb.append("\n"); 
        }

        System.out.println(sb);
    }

    public static int count_bomb(int row, int col) {

        int count = 0;
        for (int[] d : dir) {

            int nr = row + d[0];
            int nc = col + d[1];
            if (0 <= nr && nr < N && 0 <= nc && nc < N && board[nr][nc] == '*') {
                count++;
            }
        }

        return count;
    }
}